prov/verbs/src/fi_verbs.h

```diff
@@ -944,23 +964,74 @@ static inline int fi_ibv_poll_reap_unsig_cq(struct fi_ibv_ep *ep)
 	return FI_SUCCESS;
 }
 
+static inline int fi_ibv_post_send(struct fi_ibv_ep *ep, struct ibv_send_wr *wr,
+				   struct ibv_send_wr **bad_wr)
+{
+	return ibv_post_send(ep->ibv_qp, wr, bad_wr);
+}
+
+static inline int
+fi_ibv_post_send_track_credits(struct fi_ibv_ep *ep, struct ibv_send_wr *wr,
+			       struct ibv_send_wr **bad_wr)
+{
+	struct fi_ibv_cq *cq = container_of(ep->util_ep.tx_cq, struct fi_ibv_cq,
+					    util_cq);
+	int credits = (int)ofi_atomic_dec32(&cq->credits);
+
+	if (credits < 0) {
+		FI_DBG_TRACE(&fi_ibv_prov, FI_LOG_EP_DATA, "CQ credits not available,"
+		       " retry later\n");
+		ofi_atomic_inc32(&cq->credits);
+		return -FI_EAGAIN;
+	}
+	return ibv_post_send(ep->ibv_qp, wr, bad_wr);
+}
+
+static inline int
+fi_ibv_poll_cq(struct fi_ibv_cq *cq, int num_entries, struct ibv_wc *wc)
+{
+	return ibv_poll_cq(cq->cq, num_entries, wc);
+}
+
+static inline int
+fi_ibv_poll_cq_track_credits(struct fi_ibv_cq *cq, int num_entries,
+			     struct ibv_wc *wc)
+{
+	int ret = ibv_poll_cq(cq->cq, num_entries, wc);
+	if (ret > 0) {
+		switch (wc->opcode) {
```

Comment:
I think you can use if (!(wc->opcode & IBV_WC_RECV)) in place of a switch statement.

Reply:
Okay, we would avoid IBV_WC_RECV and IBV_WC_RECV_RDMA_WITH_IMM as well, assuming some new recv status won't be added in the future.

Reply:
The API is designed with the above check in mind.  So new recv opcodes would likely still work.